File: binding/include/libqsystemsemaphore.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qsystemsemaphore.html
pub const qsystemsemaphore = struct {

    /// New constructs a new QSystemSemaphore object.
    pub fn New(key: ?*anyopaque) QtC.QSystemSemaphore {
        return qtc.QSystemSemaphore_new(@ptrCast(key));
    }


    /// New2 constructs a new QSystemSemaphore object.
    pub fn New2(key: []const u8) QtC.QSystemSemaphore {
        const key_str = qtc.libqt_string{
    .len = key.len,
    .data = key.ptr,
};
return qtc.QSystemSemaphore_new2(key_str);
    }


    /// New3 constructs a new QSystemSemaphore object.
    pub fn New3(key: ?*anyopaque, initialValue: i32) QtC.QSystemSemaphore {
        return qtc.QSystemSemaphore_new3(@ptrCast(key), initialValue);
    }


    /// New4 constructs a new QSystemSemaphore object.
    pub fn New4(key: ?*anyopaque, initialValue: i32, param3: i32) QtC.QSystemSemaphore {
        return qtc.QSystemSemaphore_new4(@ptrCast(key), initialValue, @intCast(param3));
    }


    /// New5 constructs a new QSystemSemaphore object.
    pub fn New5(key: []const u8, initialValue: i32) QtC.QSystemSemaphore {
        const key_str = qtc.libqt_string{
    .len = key.len,
    .data = key.ptr,
};
return qtc.QSystemSemaphore_new5(key_str, initialValue);
    }


    /// New6 constructs a new QSystemSemaphore object.
    pub fn New6(key: []const u8, initialValue: i32, mode: i32) QtC.QSystemSemaphore {
        const key_str = qtc.libqt_string{
    .len = key.len,
    .data = key.ptr,
};
return qtc.QSystemSemaphore_new6(key_str, initialValue, @intCast(mode));
    }


    pub fn Tr(sourceText: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSystemSemaphore_Tr(@ptrCast(sourceText));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsystemsemaphore.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetNativeKey(self: ?*anyopaque, key: ?*anyopaque) void {
        qtc.QSystemSemaphore_SetNativeKey(@ptrCast(self), @ptrCast(key));
    }

    pub fn SetNativeKey2(self: ?*anyopaque, key: []const u8) void {
        const key_str = qtc.libqt_string{
    .len = key.len,
    .data = key.ptr,
};
qtc.QSystemSemaphore_SetNativeKey2(@ptrCast(self), key_str);
    }

    pub fn NativeIpcKey(self: ?*anyopaque, ) QtC.QNativeIpcKey {
        return qtc.QSystemSemaphore_NativeIpcKey(@ptrCast(self));
    }

    pub fn SetKey(self: ?*anyopaque, key: []const u8) void {
        const key_str = qtc.libqt_string{
    .len = key.len,
    .data = key.ptr,
};
qtc.QSystemSemaphore_SetKey(@ptrCast(self), key_str);
    }

    pub fn Key(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSystemSemaphore_Key(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsystemsemaphore.Key: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Acquire(self: ?*anyopaque, ) bool {
        return qtc.QSystemSemaphore_Acquire(@ptrCast(self));
    }

    pub fn Release(self: ?*anyopaque, ) bool {
        return qtc.QSystemSemaphore_Release(@ptrCast(self));
    }

    pub fn Error(self: ?*anyopaque, ) i32 {
        return qtc.QSystemSemaphore_Error(@ptrCast(self));
    }

    pub fn ErrorString(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSystemSemaphore_ErrorString(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsystemsemaphore.ErrorString: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn IsKeyTypeSupported(typeVal: u16) bool {
        return qtc.QSystemSemaphore_IsKeyTypeSupported(@intCast(typeVal));
    }

    pub fn PlatformSafeKey(key: []const u8) QtC.QNativeIpcKey {
        const key_str = qtc.libqt_string{
    .len = key.len,
    .data = key.ptr,
};
return qtc.QSystemSemaphore_PlatformSafeKey(key_str);
    }

    pub fn LegacyNativeKey(key: []const u8) QtC.QNativeIpcKey {
        const key_str = qtc.libqt_string{
    .len = key.len,
    .data = key.ptr,
};
return qtc.QSystemSemaphore_LegacyNativeKey(key_str);
    }

    pub fn Tr2(sourceText: []const u8, disambiguation: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSystemSemaphore_Tr2(@ptrCast(sourceText), @ptrCast(disambiguation));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsystemsemaphore.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(sourceText: []const u8, disambiguation: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSystemSemaphore_Tr3(@ptrCast(sourceText), @ptrCast(disambiguation), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsystemsemaphore.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetNativeKey22(self: ?*anyopaque, key: ?*anyopaque, initialValue: i32) void {
        qtc.QSystemSemaphore_SetNativeKey22(@ptrCast(self), @ptrCast(key), initialValue);
    }

    pub fn SetNativeKey3(self: ?*anyopaque, key: ?*anyopaque, initialValue: i32, param3: i32) void {
        qtc.QSystemSemaphore_SetNativeKey3(@ptrCast(self), @ptrCast(key), initialValue, @intCast(param3));
    }

    pub fn SetNativeKey23(self: ?*anyopaque, key: []const u8, initialValue: i32) void {
        const key_str = qtc.libqt_string{
    .len = key.len,
    .data = key.ptr,
};
qtc.QSystemSemaphore_SetNativeKey23(@ptrCast(self), key_str, initialValue);
    }

    pub fn SetNativeKey32(self: ?*anyopaque, key: []const u8, initialValue: i32, mode: i32) void {
        const key_str = qtc.libqt_string{
    .len = key.len,
    .data = key.ptr,
};
qtc.QSystemSemaphore_SetNativeKey32(@ptrCast(self), key_str, initialValue, @intCast(mode));
    }

    pub fn SetNativeKey4(self: ?*anyopaque, key: []const u8, initialValue: i32, mode: i32, typeVal: u16) void {
        const key_str = qtc.libqt_string{
    .len = key.len,
    .data = key.ptr,
};
qtc.QSystemSemaphore_SetNativeKey4(@ptrCast(self), key_str, initialValue, @intCast(mode), @intCast(typeVal));
    }

    pub fn SetKey2(self: ?*anyopaque, key: []const u8, initialValue: i32) void {
        const key_str = qtc.libqt_string{
    .len = key.len,
    .data = key.ptr,
};
qtc.QSystemSemaphore_SetKey2(@ptrCast(self), key_str, initialValue);
    }

    pub fn SetKey3(self: ?*anyopaque, key: []const u8, initialValue: i32, mode: i32) void {
        const key_str = qtc.libqt_string{
    .len = key.len,
    .data = key.ptr,
};
qtc.QSystemSemaphore_SetKey3(@ptrCast(self), key_str, initialValue, @intCast(mode));
    }

    pub fn Release1(self: ?*anyopaque, n: i32) bool {
        return qtc.QSystemSemaphore_Release1(@ptrCast(self), n);
    }

    pub fn PlatformSafeKey2(key: []const u8, typeVal: u16) QtC.QNativeIpcKey {
        const key_str = qtc.libqt_string{
    .len = key.len,
    .data = key.ptr,
};
return qtc.QSystemSemaphore_PlatformSafeKey2(key_str, @intCast(typeVal));
    }

    pub fn LegacyNativeKey2(key: []const u8, typeVal: u16) QtC.QNativeIpcKey {
        const key_str = qtc.libqt_string{
    .len = key.len,
    .data = key.ptr,
};
return qtc.QSystemSemaphore_LegacyNativeKey2(key_str, @intCast(typeVal));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QSystemSemaphore_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/systemsemaphore.html#types
pub const enums = struct {
    pub const AccessMode = enum {
        pub const Open: i32 = 0;
        pub const Create: i32 = 1;
    };

    pub const SystemSemaphoreError = enum {
        pub const NoError: i32 = 0;
        pub const PermissionDenied: i32 = 1;
        pub const KeyError: i32 = 2;
        pub const AlreadyExists: i32 = 3;
        pub const NotFound: i32 = 4;
        pub const OutOfResources: i32 = 5;
        pub const UnknownError: i32 = 6;
    };

};

File: binding/src/libqsystemsemaphore.cpp
#include <QNativeIpcKey>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QSystemSemaphore>
#include <qsystemsemaphore.h>
#include "libqsystemsemaphore.h"
#include "libqsystemsemaphore.hxx"

QSystemSemaphore* QSystemSemaphore_new(const QNativeIpcKey* key) {
	 return new QSystemSemaphore(*key);
}

QSystemSemaphore* QSystemSemaphore_new2(const libqt_string key) {
	 return new QSystemSemaphore(QString::fromUtf8(key.data, key.len));
}

QSystemSemaphore* QSystemSemaphore_new3(const QNativeIpcKey* key, int initialValue) {
	 return new QSystemSemaphore(*key, initialValue);
}

QSystemSemaphore* QSystemSemaphore_new4(const QNativeIpcKey* key, int initialValue, int param3) {
	 return new QSystemSemaphore(*key, initialValue, static_cast<QSystemSemaphore::AccessMode>(param3));
}

QSystemSemaphore* QSystemSemaphore_new5(const libqt_string key, int initialValue) {
	 return new QSystemSemaphore(QString::fromUtf8(key.data, key.len), initialValue);
}

QSystemSemaphore* QSystemSemaphore_new6(const libqt_string key, int initialValue, int mode) {
	 return new QSystemSemaphore(QString::fromUtf8(key.data, key.len), initialValue, static_cast<QSystemSemaphore::AccessMode>(mode));
}

libqt_string QSystemSemaphore_Tr(const char* sourceText) {
	QString _ret = QSystemSemaphore::tr(sourceText);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QSystemSemaphore_SetNativeKey(QSystemSemaphore* self, const QNativeIpcKey* key) {
	self->setNativeKey(*key);
}

void QSystemSemaphore_SetNativeKey2(QSystemSemaphore* self, const libqt_string key) {
	self->setNativeKey(QString::fromUtf8(key.data, key.len));
}

QNativeIpcKey* QSystemSemaphore_NativeIpcKey(const QSystemSemaphore* self) {
	return new QNativeIpcKey(self->nativeIpcKey());
}

void QSystemSemaphore_SetKey(QSystemSemaphore* self, const libqt_string key) {
	self->setKey(QString::fromUtf8(key.data, key.len));
}

libqt_string QSystemSemaphore_Key(const QSystemSemaphore* self) {
	QString _ret = self->key();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

bool QSystemSemaphore_Acquire(QSystemSemaphore* self) {
	return self->acquire();
}

bool QSystemSemaphore_Release(QSystemSemaphore* self) {
	return self->release();
}

int QSystemSemaphore_Error(const QSystemSemaphore* self) {
	return self->error();
}

libqt_string QSystemSemaphore_ErrorString(const QSystemSemaphore* self) {
	QString _ret = self->errorString();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

bool QSystemSemaphore_IsKeyTypeSupported(quint16 typeVal) {
	return QSystemSemaphore::isKeyTypeSupported(static_cast<QNativeIpcKey::Type>(typeVal));
}

QNativeIpcKey* QSystemSemaphore_PlatformSafeKey(const libqt_string key) {
	return new QNativeIpcKey(QSystemSemaphore::platformSafeKey(QString::fromUtf8(key.data, key.len)));
}

QNativeIpcKey* QSystemSemaphore_LegacyNativeKey(const libqt_string key) {
	return new QNativeIpcKey(QSystemSemaphore::legacyNativeKey(QString::fromUtf8(key.data, key.len)));
}

libqt_string QSystemSemaphore_Tr2(const char* sourceText, const char* disambiguation) {
	QString _ret = QSystemSemaphore::tr(sourceText, disambiguation);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QSystemSemaphore_Tr3(const char* sourceText, const char* disambiguation, int n) {
	QString _ret = QSystemSemaphore::tr(sourceText, disambiguation, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QSystemSemaphore_SetNativeKey22(QSystemSemaphore* self, const QNativeIpcKey* key, int initialValue) {
	self->setNativeKey(*key, initialValue);
}

void QSystemSemaphore_SetNativeKey3(QSystemSemaphore* self, const QNativeIpcKey* key, int initialValue, int param3) {
	self->setNativeKey(*key, initialValue, static_cast<QSystemSemaphore::AccessMode>(param3));
}

void QSystemSemaphore_SetNativeKey23(QSystemSemaphore* self, const libqt_string key, int initialValue) {
	self->setNativeKey(QString::fromUtf8(key.data, key.len), initialValue);
}

void QSystemSemaphore_SetNativeKey32(QSystemSemaphore* self, const libqt_string key, int initialValue, int mode) {
	self->setNativeKey(QString::fromUtf8(key.data, key.len), initialValue, static_cast<QSystemSemaphore::AccessMode>(mode));
}

void QSystemSemaphore_SetNativeKey4(QSystemSemaphore* self, const libqt_string key, int initialValue, int mode, quint16 typeVal) {
	self->setNativeKey(QString::fromUtf8(key.data, key.len), initialValue, static_cast<QSystemSemaphore::AccessMode>(mode), static_cast<QNativeIpcKey::Type>(typeVal));
}

void QSystemSemaphore_SetKey2(QSystemSemaphore* self, const libqt_string key, int initialValue) {
	self->setKey(QString::fromUtf8(key.data, key.len), initialValue);
}

void QSystemSemaphore_SetKey3(QSystemSemaphore* self, const libqt_string key, int initialValue, int mode) {
	self->setKey(QString::fromUtf8(key.data, key.len), initialValue, static_cast<QSystemSemaphore::AccessMode>(mode));
}

bool QSystemSemaphore_Release1(QSystemSemaphore* self, int n) {
	return self->release(n);
}

QNativeIpcKey* QSystemSemaphore_PlatformSafeKey2(const libqt_string key, quint16 typeVal) {
	return new QNativeIpcKey(QSystemSemaphore::platformSafeKey(QString::fromUtf8(key.data, key.len), static_cast<QNativeIpcKey::Type>(typeVal)));
}

QNativeIpcKey* QSystemSemaphore_LegacyNativeKey2(const libqt_string key, quint16 typeVal) {
	return new QNativeIpcKey(QSystemSemaphore::legacyNativeKey(QString::fromUtf8(key.data, key.len), static_cast<QNativeIpcKey::Type>(typeVal)));
}

void QSystemSemaphore_Delete(QSystemSemaphore* self) {
    delete self;
}

